cloudinit/cmd/daemon.py

```diff
@@ -0,0 +1,109 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+""" Run cloud-init in daemon mode. """
+
+import argparse
+import contextlib
+import json
+import os
+import socket
+import sys
+
+from cloudinit.cmd.main import main as cimain
+
+NAME = 'daemon'
+CI_SOCKET = "/run/cloud-init/unix.socket"
+
+
+@contextlib.contextmanager
+def DomainServer(addr):
+    sock = None
+    try:
+        basedir = os.path.dirname(addr)
+        if not os.path.exists(basedir):
+            os.makedirs(basedir)
+        if os.path.exists(addr):
+            os.unlink(addr)
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.bind(addr)
+        sock.listen(10)
+        yield sock
+    finally:
+        if sock:
+            sock.close()
+        if os.path.exists(addr):
+            os.unlink(addr)
+
+
+def log(stream, msg):
+    stream.write(msg + '\n')
+    stream.flush()
+
+
+def info(msg):
+    log(sys.stdout, "INFO: " + str(msg))
+
+
+def error(msg):
+    log(sys.stderr, "ERROR: " + str(msg))
+
+
+def handle_args(name, args):
+    with DomainServer(CI_SOCKET) as sock:
+        expected = set(['local', 'net', 'modules', 'final'])
+        completed = set()
+        while completed != expected:
+            conn, _ = sock.accept()
+            data = conn.recv(1024)
+            try:
+                message = json.loads(data.decode('utf-8'))
+            except Exception as e:
+                error(e)
+                message = None
+            if message:
+                command = message.get('command')
+                if not command:
+                    continue
+                if command == 'local':
+                    stage = 'local'
+                    sysv_args = ['cloud-init', 'init', '--local']
+                elif command == 'net':
+                    stage = 'net'
+                    sysv_args = ['cloud-init', 'init']
+                elif command == 'modules':
+                    stage = 'modules'
+                    sysv_args = ['cloud-init', 'modules', '--mode=config']
+                elif command == 'final':
+                    stage = 'final'
+                    sysv_args = ['cloud-init', 'modules', '--mode=final']
+
+                try:
+                    cimain(sysv_args=sysv_args)
+                except Exception:
```

Comment:
Since cimain raises a SystemExit we need to handle this here and continue otherwise the daemon exits on any successful command run. Probably also need to check the error code and log an error(). in the daemon for debug/triage maybe.

Also calling into the cli passing in sysv_args feels like we are calling in at too high a level. But, not calling into cmd/main.py:main means we'd have to do the setupBasiclogging and signal handler bits here in the daemon (or tease them out into separate functions inside cmd/main.py so they can be called from daemon and from cli separately.
I think this aspect requires a bit more discussion before we can proceed on this branch, as that may involve more work than we think is worth it for this iteration.

Reply:
Right, so the odd bit is that currently we always run through that since cloud-init is exec'ed 4 times.   I see it as an optimization if we ran some of the initial setup once in the daemon, and then never again for each of the stages.  However, there is an impact on the contents of cloud-init.log which we parse for analyze.  Already, cloud-init analyze on a daemon run produces incorrect output.  I've not yet sorted that out yet.

Reply:
I've sorted the output issue; it was related to retaining python logging configuration between stage runs.

Reply:
> Since cimain raises a SystemExit we need to handle this here and continue otherwise the daemon exits on any successful command run. Probably also need to check the error code and log an error(). in the daemon for debug/triage maybe.

cloudinit/cmd/main.py:main() does _NOT_ raise system exit;, rather  if one calls main as a module, and it's run as python's __main__, we call main(sys.argv) and then sys.exit().

daemon.py explicitly imports main.py:main() which only returns an integer precisely to avoid having the daemon exit until we want.

The intended design for daemon mode is to run  each stage as they do now; independent executions.  The default scenario (all stages succeed) daemon will invoke cimain with args (just like the __main__ function in main.py does) and after all stages have run, daemon will exit.  This allows cloud-init to behave exactly like it does today.

In the case that one or more cloud-init stages fail, the systemd units do not depend on the success of previous stages, rather they run anyhow and generally produce additional error messages that are logged.  Daemon mode, at this point, will retain this behavior; again the goal is to mirror existing behavior (we may want to modify this in the future, but for now we want to make such changes after we've landed support for this mode).

Reply:
> Also calling into the cli passing in sysv_args feels like we are calling in at too high a level. But, not calling into cmd/main.py:main means we'd have to do the setupBasiclogging and signal handler bits here in the daemon (or tease them out into separate functions inside cmd/main.py so they can be called from daemon and from cli separately.

For now, I want to retain existing behavior as much as possible (whilst also getting a bit of speed up due to reduce python startup overhead); so long term, yes, we can skip more of this setup that was done when cloud-init was 4 separate invocations after we get this landed.  This would be a further speed-up improvement;  I've captured some additional data around the time between invoking cimain() with args for each stage, and the values we get from cloud-init analyze; and there is a chunk of time that we can drop with an additional refactor that accounts for this but doesn't break analyze parsing.